xtask: Generate ELF directly

tdewey-rivos opened the pull request:
On a rv64 target, I observed that the translation from srec -> elf failed to produce a viable boot image for QEMU.

This change has xtask generate the ELF directly, rather than relying on objcopy. Observed results show the expected ELF output.

h/t Zach Reizner

Commit:
xtask: Generate ELF directly

On an rv64 targets, I observed that the translation from srec -> elf
failed to produce a viable boot image for QEMU.

This change has xtask generate the ELF directly, rather than relying on
objcopy. Observed results show the expected ELF output.

h/t Zach Reizner

Commit:
cargo.lock: Update to match xtask-elf changes

Commit:
xtask: Fix formatting

Commit:
formatting: Remove extraneous blank line

FawazTirmizi commented on build/xtask/src/dist.rs:
```diff
@@ -2462,6 +2477,94 @@ fn write_srec(
+        sections_data.extend_from_slice(sec.data.as_slice());
+    }
+
+    // HACK(zach): Bonus RAM section
+    program_headers.push(ProgramHeader {
+        p_type: PT_LOAD,
+        p_flags: PF_W | PF_R,
+        p_offset: 0,
+        p_vaddr: 0x30000000,
+        p_paddr: 0x30000000,
+        p_filesz: 0,
+        p_memsz: 0x100000,
+        p_align: 0x1000,
+    });
```
What is this hack for?

zachreizner replied:
When loading ELF files, QEMU will actually create memory segments for each program header, unlike an actual bootloader which obviously can't create memory out of thin air. This hack takes advantage of that fact to create a segment of read-write memory for use by Hubris.

FawazTirmizi replied:
What does Hubris use it for?

zachreizner replied:
The app used that memory for its RAM area. At the time, I was copying one of the hifive board's linker layouts which needed some place to put the RAM. The board would of course already have the RAM there, but QEMU needed to have some extra input to make it work. Doing it via the program headers is convenient, but it's just a hack. A better way would be to have a specific board built-in to QEMU that already has the memory it needs regardless of the program headers.

adurbin-rivos replied:
Could one point to what QEMU is doing so I can follow the logic? I would expect QEMU to have a notion of physical address space that programs can be loaded into -- not having the program dictate physical address space addressability.

beezow replied:
Is this an alternative to adding a ram device to the machine definition in the Qemu source?

tdewey-rivos replied:
@adurbin-rivos The call chain gets interesting from here: https://github.com/qemu/qemu/blob/b3cd3b5a66f0dddfe3d5ba2bef13cd4f5b89cde9/hw/core/loader.c#L449, which then calls https://github.com/qemu/qemu/blob/b3cd3b5a66f0dddfe3d5ba2bef13cd4f5b89cde9/include/hw/elf_ops.h#L315

It's not the most straightforward C, but around like 562 in elf_ops.h you can see that the file_size < mem_size check will succeed, and QEMU will consequently zero the address specified (in this case, the whole section-address region).

tdewey-rivos replied:
@beezow Yes - it's an elegant hack that gets some address space zero'd ready for your use. The alternative would, I guess, be to modify the Qemu board to set up some memory at the a priori known address, but I'm not convinced that's a better solution.

tdewey-rivos replied:
Removed this block, as it fouls the QEMU ELF loader (which for RISC-V appears to deliberately ignore the elf start address, and picks the lowest address).

furquan-rivos commented on build/xtask/Cargo.toml:
```diff
@@ -19,7 +19,7 @@ toml = "0.5.6"
 ron = "0.7"
```
General comment on the commit message: I think we would want to do the same thing for rv32 as well. Probably we should land this in `rivos/main` and start using it for both rv32 and rv64. Thoughts? @FawazTirmizi @beezow

FawazTirmizi replied:
I agree, especially if we want to change the build flow to go from ELF -> {bin, ihex, srec}.

tdewey-rivos replied:
ACK on rv32.

I'm not convinced that we actually want to use ELF as the 'reference' format, however, given we're adding additional program headers during elf generation.

My preference would be that we treat ELF as a debug target for QEMU, and continue with binary generation in parallel. Like two parallel trees of image derivation, depending on desired use.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -431,6 +431,14 @@ pub fn package(
 
         translate_srec_to_other_formats(&cfg, image_name, "combined")?;
 
+        if cfg.toml.target.contains("riscv") {
```
The way the build flow happens with these changes is:
1. Build srec image
2. Translate srec to other formats (bin, ihex) for rv and (bin, ihex, elf) for !rv
3. Build elf image for rv

I think we should instead create a consistent flow for all architectures:
1. Build elf image
2. Translate to all other formats (ihex, bin, srec)

That way we won't have to maintain different paths here for different architectures. Is there any limitation currently because of which we cannot do that?

adurbin-rivos replied:
None that I'm aware of. The original ELF has all the interesting information to form the formats that drop said infofmaiton.

tdewey-rivos replied:
The only sticking point I'm aware of is the additional ram-section, which we certainly don't want in binaries aimed elsewhere.

I took this approach from a position of cautious change - in particular, I didn't want to change this for non-rv targets, as their Qemu HW might take a different approach.

I agree in principle, however, that it would be better to have a consistent flow (or a 'true' federated image generation, where the app.toml has a mechanism to specify image generation instructions).

tdewey-rivos replied:
All targets will now use the manual ELF generation. There's one more change coming to not break ARM, et al, but the groundwork is in.

furquan-rivos replied:
> The only sticking point I'm aware of is the additional ram-section, which we certainly don't want in binaries aimed elsewhere.

I don't think we need it even for our own builds. Based on @zachreizner's comment, it looks like a hack that was done for a specific case. I see that you decided to drop the hack in the latest commit.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
+        // the file (what the loop adjusts) with its location in virtual address space (which was
+        // fixed by the compiler already).
+        while ((base as usize) & 0xfff) != (sections_data.len() & 0xfff) {
+            sections_data.push(0);
```
Should it be filling with 0x0 or 0xff?

adurbin-rivos replied:
I think the 'sections' only contain loadable segments? I'm not sure the padding value matters in practice aside from on-flash usage (which I thin is why you're asking). Whichever value we choose we should note in commentary there was thought put into the reason for the value chosen (and the alternatives).

furquan-rivos replied:
Yes, I was thinking about the flash usage. If we are adding some kind of padding to align the sections, then having `0xff` just makes it easier to skip the write on flash since it erases default to 0xff.

I agree that we should add comments around the reason behind some of the choices so that it is helpful later on.

tdewey-rivos replied:
Adjusted to fill with 0xFF, added a comment along @furquan-rivos's reasoning.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
+
+        program_headers.push(ProgramHeader {
+            p_type: PT_LOAD,
+            p_flags: PF_X | PF_W | PF_R,
```
I had raised this [bug](http://go/ji/SW-474) to track how we want to set the flags differently for different regions.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
     Ok(())
 }
 
+fn write_elf(
```
We will probably need write_elf32 and write_elf64 if we have to support consistent flows for all architectures.

beezow replied:
I feel like it should be able to share most of the code.  From what I have seen, alot of the `goblin::elf` apis can be agnostic to if it is 32 or 64 bit.

tdewey-rivos replied:
Agree w/ @beezow, I suspect this will be relatively straightforward.

tdewey-rivos replied:
Changed in latest revision.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
+            p_paddr: base as u64,
+            p_filesz: sec.data.len() as u64,
+            p_memsz: sec.data.len() as u64,
+            p_align: 0x1000,
```
Why 4KiB?

adurbin-rivos replied:
It should follow alignment of the base address.

tdewey-rivos replied:
This block was taken from the original patch - but looking at the alignment of base it is only actually 16 byte aligned. Suggest this comes in to a later PR, as I'm making changes to the ELF writing later to include section headers (without which, objcopy will _not_ convert to other formats)

furquan-rivos replied:
Ack

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
+        p_type: PT_LOAD,
+        p_flags: PF_W | PF_R,
+        p_offset: 0,
+        p_vaddr: 0x30000000,
```
This address might not be RAM depending upon the machine model.

tdewey-rivos replied:
True. Ideally, we'd have this provided by board configuration, or app.toml

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
+        .iter_mut()
+        .for_each(|phdr| phdr.p_offset += sections_data_offset as u64);
+
+    let mut elf_out = vec![0; sections_data_offset + sections_data.len()];
```
Should this be 0xff instead of 0? That is what the flash erase value would be.

furquan-rivos commented:
> [furquan-rivos](https://github.com/furquan-rivos) requested review from [adurbin-rivos](https://github.com/adurbin-rivos), [ted-logan](https://github.com/ted-logan) and https://github.com/orgs/rivosinc/teams/hubris-humility-maintainers [37 seconds ago](https://github.com/rivosinc/hubris/pull/45#event-7527426574)

Woops. I gave the wrong PR# to my script and it ended up adding `ted-logan` on this PR again which caused your earlier approval to disappear :(. Sorry about that. It is definitely a weird side-effect.

adurbin-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
+        // Incredibly simple algorithm that attempts to align where this program section appears in
+        // the file (what the loop adjusts) with its location in virtual address space (which was
+        // fixed by the compiler already).
+        while ((base as usize) & 0xfff) != (sections_data.len() & 0xfff) {
```
What is base? VA? Load address? 

It's seems we're really trying to build a loadable segment within a file.  Only when one is mmap()ing a file for loading is it necessary to match the offsets. It seems that's a requirement for QEMU purposes? However, that's not necessary if one is copying data to the loadable address. But to meet the strictest requirements we're using 4KiB as an atomic unit to calculate modulo offsets. Given that (which would be helpful to have in commentary), the program headers live at a particular place in the file on disk. So we should really be starting our offset comparison based on the size the resultant program headers occupy, imo.

Are we assuming all VAs are in increasing order? We may want to sort the sections by base as well.

furquan-rivos replied:
> What is base? VA? Load address?

base is PA i.e. load address.

> It's seems we're really trying to build a loadable segment within a file. Only when one is mmap()ing a file for loading is it necessary to match the offsets. It seems that's a requirement for QEMU purposes?

I think its a combination of:
1. QEMU requirement for using generic loader which assumes the combined image (kernel + apps) are located in the memory at certain address. I think for ELF qemu should do the right thing for loading based on ELF headers.
2. Hubris assuming that we run code from flash. So, there is no loading currently for any of the code/rodata sections.

> But to meet the strictest requirements we're using 4KiB as an atomic unit to calculate modulo offsets.

I am not sure if we really need to do this. There is already logic in xtask elsewhere to ensure that we  align the task memory regions. See `task_memory_alignment`.

> Are we assuming all VAs are in increasing order? We may want to sort the sections by base as well.

Sections are stored in a `BTreeMap` which is a map sorted using the keys as base.

adurbin-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
+        program_headers.push(ProgramHeader {
+            p_type: PT_LOAD,
+            p_flags: PF_X | PF_W | PF_R,
+            p_offset: sections_data.len() as u64,
```
I don't see how we're accounting for program header length by using sections_data.len(). If the initial section is aligned properly (e.g. 4KiB alignment with sections_data.len() == 0), then we'll be colliding with the program headers.

furquan-rivos replied:
[Lines 2216-2221](https://github.com/rivosinc/hubris/pull/45/files/464da40f2a100382461b84037de1ce5918016706#diff-8cae0109f2f633af27f329466195e57580c67db93f652ca33e917d685104ba46R2216) later adjust the p_offset for each program header.

Commit:
xtask: Restrict manual ELF generation to rv64

The prior checks were too broad, and may have produced an ELF64
for rv32 targets.

Signed-off-by: Tom Dewey <tdewey@rivosinc.com>

Commit:
xtask: Remove RAM section hack

This creates a particularly troublsome interaction with QEMUs elf
loader, which explicitly selects the _lowest address section_ as
the jump point for kernels (claiming this works to support FreeBSD).

So for now, undo the RAM allocation hack, allowing us to boot
without modifying QEMU or using the generic loader.

Signed-off-by: Tom Dewey <tdewey@rivosinc.com>

Commit:
xtask: Generate all ELF manually

Use the objcopy target to identify if we're generating an ELF32 or
ELF64, and pass that state to write_elf, which then uses it to map
to a Goblin container size.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -446,13 +446,16 @@ pub fn package(
 
         translate_srec_to_other_formats(&cfg, image_name, "combined")?;
```
I think the `write_srec` can now be dropped and `translate_srec_to_other_formats` can instead do `translate_elf_to_other_formats` including ihex, bin and srec. Not sure if that is the upcoming change you commented about in the other thread.

tdewey-rivos replied:
I don't believe we can outright drop translate_sec* until we have a better story for the signed code path - I think that would put us on a path to interesting clashes with upstream that I don't think we want.

A follow-up PR will have all formats initially generated from the elf - but retaining translate_srec for the signing path. I'm still verifying objcopy is generating correct images - there's an appreciable size difference in ihex & srec.

furquan-rivos replied:
I am confused. Isn't `translate_srec_to_other_formats` only doing objcopy from srec to other formats? There is no signing involved within that function. Do you mean that srec images generated by `write_srec` are different than the ones generated by elf->srec conversion? I think that's what you mean based on your last statement, but just want to be sure.

tdewey-rivos replied:
Correct - it only converts, but in two different flows. In the unsigned image flow, `translate_srec...` is only used to derive the alternative image formats - but in the signed image flow (line 467 onwards), Hubris signs the derived binary, creating a `combined_rsa.bin`, which is then re-converted _back_ to srec (as `final.srec`) before yet _another_ set of derived images (`final.x`) is then derived from that signed srec.

I'm not sure we want to make a call on image signing as yet - so my plan was to leave `translate_srec...` and `write_srec` in place, until we can decide what to do about image signing.

furquan-rivos replied:
Understood about the signing flow. My point was that the common flow can be updated to do:
1. Generate elf
2. Translate elf to other formats

If signing is required, then it would do the `binary_to_srec` and `translate_srec_to_other_formats`. That way you won't need `write_srec` at all and the srec handling would be limited to only the signing path which as you mentioned we don't really want to care about right now. That way, we won't have any special srec generation code in our common path (just less things to worry about from image management standpoint).

Anyways, if you want to retain the srec path, I think it would be good to at least add a comment in the file regarding why it is left around and probably a JIRA issue to track this as a TODO for later.

tdewey-rivos replied:
Ah, so there's a bit more to unpack here:

1. This MR cannot derive other formats from the elf. objcopy requires section headers in addition to program headers, even though QEMU is happy enough without.
2. In the patch that's immediately after this, the derived srec is not identical, and I don't yet have a mechanism to confirm if that deviation is material. While it matches in terms of overall size & record count, the records themselves have different header data. Comparing the derived files suggests that objcopy is reordering blocks.

Commit:
xtask: Fixup ARM ELF Generation

Prior to this change, all generated ELF files were marked for
RISC-V. Other platforms still exist, so we should make sure that
they can also have generated ELF files.

Commit:
xtask: Derive all other images from ELF

Change how we generate ELF images, such that these can now be consumed
by objcopy. To do this, introduce section headers.

Additionally, fix up how we generate the binary image itself. Rather
than growing the result vector, instead pre-allocate the expected size,
fill it with 0xFF, and then in-place-replace with the section data.

Confirmed building rv64-qemu-virt that I see identical srec images
generated.

ghost commented:
I believe the latest changes address all outstanding comments - we retain the via-srec signing path, but only the bits that we strictly require. Alignments now match the rest of the Hubris build system, and on-disk layout matches previous output for binary & srec formats (ihex sees some difference that appears to be related only to the gap-filling). Introducing ELF section headers unlocks the objcopy path, so ELF becomes the canonical format.

furquan-rivos commented:
One of the jobs failed with this change:

https://github.com/rivosinc/hubris/actions/runs/3219763270/jobs/5265575281

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2502,7 +2504,8 @@ fn write_elf(
+            sections_data.push(0xff);
         }
 
         program_headers.push(ProgramHeader {
```
Since this function handles both 32-bit and 64-bit architectures, won't the program header fields have to be handled accordingly to be either 4- or 8-byte long? e.g. vaddr, paddr, filesz, etc.

tdewey-rivos replied:
It's not exceptionally clear how this is safe, I agree.

The goblin documentation suggests that these unadorned types are 'universal' types, and that they will be scaled according to the Container size we set in the ctx above. I can't say I've found where it's doing that, exactly, though.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -438,21 +438,15 @@ pub fn package(
 
         // Generate combined SREC, which is our source of truth for combined images.
```
Comment needs update.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2512,10 +2500,17 @@ fn write_elf(
             p_filesz: sec.data.len() as u64,
             p_memsz: sec.data.len() as u64,
-            p_align: 0x1000,
+            p_align: 0x20, // This matches the alignment guarantees of the kernel & task build
```
Is this coming from the max-page-size and common-page-size in `link` and `build`?

tdewey-rivos replied:
Yes, I believe so. For future work, we could explore changing all these alignments to be a target-specific property (likely as part of a wider refactor of xtask - the signing flow in particular is extremely device-specific)

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2524,28 +2519,74 @@ fn write_elf(
+    let shstrtab_offset =
+        (sections_data_offset + sections_data.len() + 0xfff) & !0xfff;
+
+    // Create a Section Header String Table, to hold the actual section
+    // names.
```
Would the section names hold the task names?

tdewey-rivos replied:
I don't think we have access to that during this point of packaging - but we _can_ use the source filename, which I believe includes the end 'elf' (which is typically a 1:1 mapping between source elf and task, and a many:1 relationship between sections and source elf).

furquan-rivos replied:
Sounds good. We can also do this as a follow up.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2524,28 +2519,74 @@ fn write_elf(
+    let mut null_section_header = SectionHeader::new();
+    null_section_header.sh_type = SHT_NULL;
+
+    // We create a fake '.text' section to guide objcopy toward the instruction
+    // bitstream.
```
Why not create a separate section for each task and kernel? That would be similar to how the srec->elf translation worked before. Example:

```
Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .sec1             PROGBITS         0000000090000400  00000040
       0000000000010400  0000000000000000  WA       0     0     1
  [ 2] .sec2             PROGBITS         0000000090010800  00010440
       000000000000c900  0000000000000000  WA       0     0     1
  [ 3] .sec3             PROGBITS         000000009001d100  0001cd40
       0000000000001700  0000000000000000  WA       0     0     1
  [ 4] .sec4             PROGBITS         000000009001e800  0001e440
       0000000000002000  0000000000000000  WA       0     0     1
  [ 5] .sec5             PROGBITS         0000000090020800  00020440
       0000000000002e40  0000000000000000  WA       0     0     1
  [ 6] .sec6             PROGBITS         0000000090023640  00023280
       00000000000021c0  0000000000000000  WA       0     0     1
  [ 7] .sec7             PROGBITS         0000000090025800  00025440
       00000000000023c0  0000000000000000  WA       0     0     1
  [ 8] .sec8             PROGBITS         0000000090027bc0  00027800
       0000000000000a28  0000000000000000  WA       0     0     1
  [ 9] .shstrtab         STRTAB           0000000000000000  00028228
       000000000000003b  0000000000000000           0     0     1
```

tdewey-rivos replied:
It's an option - happy to go down that path if we think it's preferable. Will want to test the output of objcopy after it's done.

furquan-rivos replied:
I think that could make things slightly simpler to debug when inspecting an ELF image.

tdewey-rivos replied:
Added in lasted change.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2524,28 +2519,74 @@ fn write_elf(
+    // bitstream.
+    let mut text_section_header = SectionHeader::new();
+    text_section_header.sh_type = SHT_PROGBITS;
+    text_section_header.sh_flags = (SHF_ALLOC | SHF_EXECINSTR) as u64;
```
Should this be marked as execute? The entire section is not really execute.

tdewey-rivos replied:
At this point of packaging, we don't really have the ability to distinguish between sections to be loaded or not. My understanding of the ELF documentation is that the .text section mandates SHF_ALLOC & SHF_EXECINSTR: https://man7.org/linux/man-pages/man5/elf.5.html

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2575,6 +2626,8 @@ fn objcopy_translate_format(
         .arg(out_format)
         .arg("--gap-fill")
         .arg("0xFF")
+        .arg("--srec-forceS3") // Manually constructed Srecords use the S3 format
+        .arg("--srec-len=255") // Objcopy will select a shorter line length if allowed, this forces it to match the manual Srecord construction.
```
Do these arguments need to be conditionally added if the out_format is srec?

tdewey-rivos replied:
No - objcopy ignores them entirely for non-srec outputs.

ghost commented:
hifive1-revb failed because I failed to account for the packaging tool being handed RAM sections in addition to program sections. I'm going to prune any section of `len() == 0` to mitigate this, which we previously got for free as part of the program header assembly loop. I believe pre-allocating the section data structure is still preferable, as it becomes simpler to reason about the behaviour of section emplacement.

ghost commented:
While the section sizes were incorrect on hifive1-revb, it appears that the Goblin APIs as used are not automatically converting between ELF64/ELF32. I'm going to introduce separate header construction functions for 32/64.

Commit:
elf: Fixup ELF32 generation

Prior to this commit, we inadvertently treated all ELF files as if they
were ELF64, in spite of using ELF32 sizes for assembly.

This produced predictably poor results.

As a result, this changes duplicates the ELF writing function into two
subroutines - one for ELF32, one for ELF64. The top-level 'write_elf'
becomes a facade, selecting the appropriate writing function.

Additionally, put each task code section into it's own elf section, for
clarity.

Tested on demo-rv64-qemu-virt, demo-hifive1-revb:
- readelf confirms expected header, sections
- objdump confirms sections contain expected data
- build produces all derived images with plausible content

Commit:
xtask: Unified ELF32/64 Creation

In this patch, I make liberal use of rust's macro system and the
paste crate to create macros to generate both ELF32 and ELF64 headers.

In either case, we create both sets of containers (two vectors,
shared ELF header structure, shared section header string table).
However, only one of the [32|64] vector sets is filled, according
to the type of ELF to be generated. This is controlled by inspecting
arch_constants to discover if the end elf target is elf64.

Commit:
xtask: Fixup section/end calcs, write ELF32

Capture a missing line to actually write the ELF32 file, and ensure all
section length calculations are done in AbiSize containers.

Commit:
xtask: Fixup formatting on new start/end calcs

ghost commented:
Pending any blocking comments, I believe this is now Good To Go into the tree, and has addressed the major comments of previous reviews. I've tested building rv32 & rv64, and manually inspected the output. CI appears to be passing. One potential open question is if we want to retain the macros that create the elf structures in dist.rs: I have chosen to do so to keep proximity to the call site.

furquan-rivos commented:
> Pending any blocking comments, I believe this is now Good To Go into the tree, and has addressed the major comments of previous reviews. I've tested building rv32 & rv64, and manually inspected the output. CI appears to be passing. One potential open question is if we want to retain the macros that create the elf structures in dist.rs: I have chosen to do so to keep proximity to the call site.

I haven't looked through the changes yet, but did a quick checkout of the PR and built an image for tests-hifive-inventor and tests-rv64-qemu-virt and both fail to boot on qemu with this PR. Going back to main-rv64 branch fixes the issue. Reporting this early so that you get a chance to look at it while the PR is being reviewed. Thanks!

furquan-rivos commented:
> > Pending any blocking comments, I believe this is now Good To Go into the tree, and has addressed the major comments of previous reviews. I've tested building rv32 & rv64, and manually inspected the output. CI appears to be passing. One potential open question is if we want to retain the macros that create the elf structures in dist.rs: I have chosen to do so to keep proximity to the call site.
> 
> I haven't looked through the changes yet, but did a quick checkout of the PR and built an image for tests-hifive-inventor and tests-rv64-qemu-virt and both fail to boot on qemu with this PR. Going back to main-rv64 branch fixes the issue. Reporting this early so that you get a chance to look at it while the PR is being reviewed. Thanks!

Tried both final.bin and final.elf images and none worked.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2452,69 +2452,125 @@ fn write_elf(
+        if candidate_section.1.data.len() > 0 {
+            if *candidate_section.0 < sections_base_address {
+                sections_base_address = *candidate_section.0;
+            }
```
Is this handling the condition where some task is placed before the kernel in the layout?

tdewey-rivos replied:
Yes, it's a preventative guard. I can't say I've seen an existing target that does this, but I also couldn't identify what would stop someone from putting something in advance of the kernel.

furquan-rivos commented on build/xtask/src/dist.rs:
```diff
@@ -2452,69 +2452,125 @@ fn write_elf(
+
+        let task_name = sec.source_file.file_name().to_owned().unwrap();
+
+        let mut section_name: String = String::from(".text.");
```
Just noting: Even though this adds the prefix `.text` to every section, not every section is really .text. But probably, we don't have the information at this point to determine whether it is .text or something else.

tdewey-rivos replied:
Exactly that - by the time we get the section data here, we've only really got knowledge of the bitstream, source file, and offset. I suspect we might be able to do better if we expanded the elf reading prior to the packaging step, but until I see that producing a concrete issue I'm reticent to go down that path.

ghost commented:
ACK'd QEMU boot. Will reproduce this today and keep digging.

ghost commented:
After comparing against pre-MR output in the binary file, I can see that ObjCopy is producing a very different binary when I introduce per-program section headers. Notably, it appears to prepend the binary with 0x2e0 of unexpected data - the start of which appears to be an ELF header.

Dropping the per-section Section Header, I see binary output return to the reference form. There remains some difference in the ihex & srec, but these appear to be related to objcopy filling-in gaps (I see a lot of introduced 0xFF sections).

Using the single-SectionHeader variant, I am able to boot rv64-qemu-virt in QEMU, as I would expect. I'll add another commit to this MR, where only a single section is declared.

Commit:
xtask: Use single section header for resultant image

Producing a section header per program section resulted in objcopy
reordering sections and introducing data from unexpected sources.
The resultant images were unbootable.

Return to using a single section header to encompass the entire
resultant image. Disassembly confirms this produces output that we
would expect.

Commit:
xtask: Fixup .text label offset

Move the write of the .text label into the Section Header String
Table to after the offset is captured for the reference, such
that the captured offset points at the start of the desired label,
and not the next one.

furquan-rivos reviewed (commented):
Latest change looks fine. As we discussed last week, all this effort is primarily to enable elf booting in qemu. We need to also think about the bigger picture for real platform targets and what our building and stitching should look like for that. That way we would be able to better align efforts within this and all other firmware components to build and stitch images for different "targets". I have started putting together some thoughts on this and will share with the wider team in coming weeks.

For this change, let's just give other reviewers a chance in case there are any other comments. If not, I can hit approve end of day today.

furquan-rivos reviewed (approved):
No new comments/concerns in past 1 day. So, hitting approve. P.S. this is mainly being done to make qemu happy. We will have to follow up with more changes as we chalk out details around entire firmware building and stitching.

Commit:
Merge branch 'rivos/main-rv64' into dev/tdewey/xtask-elf
